### PR TITLE
[CLOUDTRUST-4985] Fix bug to avoid removing first/last names when not expected

### PR DESCRIPTION
**Pull request by fperot74**

### CODE REVIEW

File: api/account/api.go [hunk 1]
```diff
@@ -181,23 +181,39 @@ func convertToAccreditations(ctx context.Context, values []string, logger keyclo
...
+	//userRep.Email = defaultStringPtr(user.Email, userRep.Email)
+	userRep.FirstName = defaultStringPtr(user.FirstName, userRep.FirstName)
+	userRep.LastName = defaultStringPtr(user.LastName, userRep.LastName)
 
 	var attributes = make(kc.Attributes)
-	attributes.SetStringWhenNotNil(constants.AttrbPhoneNumber, user.PhoneNumber)
+	if userRep.Attributes != nil {
+		attributes = *userRep.Attributes
+	}
+	// Don't set email as email and phoneNumber have specific validation rules
```
Comment (harture) [inline api/account/api.go:203]: Don't set phone number

File: api/account/api.go [hunk 2]
```diff
@@ -181,23 +181,39 @@ func convertToAccreditations(ctx context.Context, values []string, logger keyclo
...
+}
 
-	userRep.Username = user.Username
-	userRep.Email = user.Email
-	userRep.FirstName = user.FirstName
-	userRep.LastName = user.LastName
+// MergeUser merge new values into existing KC user representation
+func MergeUser(userRep *kc.UserRepresentation, user UpdatableAccountRepresentation) {
+	userRep.Username = defaultStringPtr(user.Username, userRep.Username)
+	// Don't set email as email and phoneNumber have specific validation rules
```
Comment (harture) [inline api/account/api.go:194]: Ok, but where do you manage it now ?

File: api/account/api.go (hunk 1 above)
Comment (harture) [reply api/account/api.go:203]: Where is it managed ?

File: api/account/api.go [hunk 3]
```diff
@@ -181,23 +181,39 @@ func convertToAccreditations(ctx context.Context, values []string, logger keyclo
...
+	//userRep.Email = defaultStringPtr(user.Email, userRep.Email)
+	userRep.FirstName = defaultStringPtr(user.FirstName, userRep.FirstName)
+	userRep.LastName = defaultStringPtr(user.LastName, userRep.LastName)
 
 	var attributes = make(kc.Attributes)
```
Comment (harture) [inline api/account/api.go:199]: Why we don't use the duplicateAttribute method ?

File: api/management/api.go [hunk 4]
```diff
@@ -298,15 +298,23 @@ type IdentityProviderRepresentation struct {
 type RequiredAction string
 
 func defaultString(actual *string, defaultValue string) *string {
```
Comment (harture) [inline api/management/api.go:300]: This method is a bit strange in its behavior. The output might be a reference to the initial string or not. I have a bad feeling about it.

**Review by harture — CHANGES_REQUESTED** [review]

File: api/account/api.go (hunk 1 above)
Comment (fperot74) [reply api/account/api.go:203]: > Don't set phone number

It is not set...

**Comment by harture** [comment]
Yes. MY comment was to update your comment. It talks about email instead of phone number 😉

File: api/account/api.go (hunk 3 above)
Comment (fperot74) [reply api/account/api.go:199]: Because it does not exist... and we are merging values, not duplicating them

File: api/management/api.go (hunk 4 above)
Comment (fperot74) [reply api/management/api.go:300]: That's the same defaultString mechanism as the one in Apache common lang 3

File: api/account/api.go (hunk 2 above)
Comment (fperot74) [reply api/account/api.go:194]: It is the responsibility of the caller (UpdateAccount in pkg/account)

Comment (basbeu) [reply api/account/api.go:194]: I do not understand the link between specific validation rules and that we do not merge them ! 
In my point of view a merge method should merge every fields.

Comment (basbeu) [reply api/account/api.go:194]: In my point of view it is really not a good idea !

**Review by basbeu — CHANGES_REQUESTED** [review]

**Comment by basbeu** [comment]
I saw that the PR is functional, but I think that the email and phone number should be merged !

File: api/account/api.go (hunk 2 above)
Comment (fperot74) [reply api/account/api.go]: Repushed with renamed function

**Review by basbeu — APPROVED** [review]

**Review by harture — APPROVED** [review]

**Pull request merged**